File: src/util/DataGenerator.java
package util;

import interfaces.IDataGenerator;

import java.util.Random;

public class DataGenerator implements IDataGenerator {
    private final Random random;

    public DataGenerator() {
        random = new Random();
    }

    @Override
    public int[] generateRandomArray(int size) {
        int[] array = new int[size];
        for (int i = 0; i < size; i++) {
            array[i] = random.nextInt(size * 10);
        }
        return array;
    }

    @Override
    public int[] generateSortedArray(int size) {
        int[] array = new int[size];
        for (int i = 0; i < size; i++) {
            array[i] = i;
        }
        return array;
    }

    @Override
    public int[] generateNearlySortedArray(int size) {
        int[] array = generateSortedArray(size);

        int elementsToShuffle = size / 20;
        for (int i = 0; i < elementsToShuffle; i++) {
            int pos1 = random.nextInt(size);
            int pos2 = random.nextInt(size);

            int temp = array[pos1];
            array[pos1] = array[pos2];
            array[pos2] = temp;
        }
        return array;
    }

    @Override
    public int[] generateReverseArray(int size) {
        int[] array = new int[size];
        for (int i = 0; i < size; i++) {
            array[i] = size - i - 1;
        }
        return array;
    }
}

File: python/data_generator.py
from idata_generator import IDataGenerator
import random

class DataGenerator(IDataGenerator):
    def generate_random_array(self, size):
        return [random.randint(0, size * 10) for _ in range(size)]

    def generate_sorted_array(self, size):
        return list(range(size))

    def generate_nearly_sorted_array(self, size):
        array = self.generate_sorted_array(size)
        elements_to_shuffle = size // 20
        for _ in range(elements_to_shuffle):
            pos1 = random.randint(0, size - 1)
            pos2 = random.randint(0, size - 1)
            array[pos1], array[pos2] = array[pos2], array[pos1]
        return array

    def generate_reverse_array(self, size):
        return list(range(size, 0, -1))